tests/system/test_pandas.py

```diff
@@ -1304,6 +1304,32 @@ def test_upload_time_and_datetime_56(bigquery_client, dataset_id):
     ]
 
 
+def test_to_dataframe_query_with_empty_results(bigquery_client):
+    """
+    JSON regression test for https://github.com/googleapis/python-bigquery/issues/1580.
+    """
+    job = bigquery_client.query(
+        """
+        select
+        123 as int_col,
+        '' as string_col,
+        to_json('{}') as json_col,
+        struct(to_json('[]') as json_field, -1 as int_field) as struct_col,
+        [to_json('null')] as json_array_col,
+        from unnest([])
+        """
+    )
+    df = job.to_dataframe()
+    assert list(df.columns) == [
+        "int_col",
+        "string_col",
+        "json_col",
+        "struct_col",
+        "json_array_col",
+    ]
+    assert len(df.index) == 0
```

Comment:
#QUESTION:

The test suite for `test_to_arrow_query_with_empty_results` is more robust than this one.
Is there a reason for the difference?

Reply:
We're using `object` dtype in pandas for `STRUCT` and `ARRAY` columns right now. This means there's not much we can inspect about subfields/subtypes like we can with Arrow.

Related: I'd like to introduce a `dtype_backend` parameter like pandas has to pandas-gbq: https://github.com/googleapis/python-bigquery-pandas/issues/621

In the meantime, I tend to use:

```python
df = (
    results
    .to_arrow(bqstorage_client=bqstorage_client)
    .to_pandas(types_mapper=lambda type_ : pandas.ArrowDtype(type_))
)
```

instead of `to_dataframe()` in my personal projects, as this will give me the more structured Arrow types all the time.